api/envoy/config/route/v3/route_components.proto

```diff
@@ -764,10 +764,106 @@ message RouteAction {
   }
 
+  // Indicates that some headers received from upstream cluster need to be rewritten. When upstream
+  // returns a header, the header will be overwritten with the given value before being sent back as
+  // response to the client.
+  //
+  // In the below example we rewrite the location header received from the upstream cluster. In the
+  // case of the location header, two things need to match before a value substitution happens for
+  // the location header - Location header's value, and a few other headers present in upstream
+  // response.
+  //
+  // If a header is specified in response_headers_to_rewrite section but is not returned from the
+  // upstream it it simply ignored.
+  //
+  //   .. code-block:: yaml
+  //
+  //     - match:
+  //         prefix: "/itm/"
+  //       route:
+  //         response_headers_to_rewrite:
+  //           "location":
+  //             header_rewrite_literal:
+  //               string_match:
+  //                 prefix: /itm/not-found/
+  //               substitution: /n/error
+  //             request_match_criteria:
+  //               and_filter:
+  //                 filters:
+  //                 - and_filter:
+  //                   filters:
+  //                   - header:
+  //                     key: content-type
+  //                     value: text/html
+  //                   - header:
+  //                     key: cookie
+  //                     value: pagetype=123
+  //                 - or_filter:
+  //                   filters:
+  //                   - header:
+  //                     key: status
+  //                     value: 301
+  //                   - header:
+  //                     key: status
+  //                     value: 302
+  //                   - header:
+  //                     key: status
+  //                     value: 307
+
+  message HeaderRewriteLiteral {
+    // String match can use prefix match and present match. In case of prefix match, the matched prefix part of
+    // the header's value is substituted. In case of present match the whole value of the header is overwritten.
+    type.matcher.v3.StringMatcher string_match = 1 [(validate.rules).message = {required: true}];
+
+    // The matched part will be replaced by the value provided in substitution.
+    string substitution = 2 [(validate.rules).string = {min_bytes: 1}];
+  }
+
+  message HeaderRewriter {
+    oneof header_rewriter_type {
+      // Describes the pattern to look for in the Header's value and what to subsitute in place of such value
+      HeaderRewriteLiteral header_rewrite_literal = 1;
+    }
+
+    // Describes which requests should be eligible for rewriting headers received from upstream. If this match
+    // criteria fails, no action would be taken on the upstream headers before sending to the end user.
+    HeaderMatcherFilter request_match_criteria = 3 [(validate.rules).message = {required: true}];
+  }
+
+  message HeaderMatcherFilter {
```

Comment:
Thanks. How strongly do @envoyproxy/api-shepherds feel about trying to clean this up further in this PR? I think the ideal would be to migrate _both_ the matching going on in `AccessLogfilter` and `HeaderMatcher` to `envoy.type.matcher.v3`.

Reply:
Sorry @htuch what are you proposing? cc @snowp @yangminzhu for other efforts in flight to merge matching behavior. I agree at a high level that it pains me to see yet another and/or, etc. in the API here.